File: Homework1/Hw1/Parser.cs
using System.Security.Cryptography.X509Certificates;

namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args, 
        out double val1, 
        out CalculatorOperation operation, 
        out double val2)
    {
        if (IsArgLengthSupported(args))
        {
            
            bool tryVal1 = double.TryParse(args[0], out val1);
            bool tryVal2 = double.TryParse(args[2], out val2);
            if (tryVal1 && tryVal2)
                operation = ParseOperation(args[1]);
            else
                throw new ArgumentException();
        }
        else
        {
            throw new ArgumentException();
        }
    }

    private static bool IsArgLengthSupported(string[] args) => args.Length == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        if (arg == "+")
            return CalculatorOperation.Plus;
        else if (arg == "-")
            return CalculatorOperation.Minus;
        else if (arg == "*")
            return CalculatorOperation.Multiply;
        else if (arg == "/")
            return CalculatorOperation.Divide;
        else
            throw new InvalidOperationException();
    }
}

File: Homework5/Hw5/Parser.fs
module Hw5.Parser

open System
open Hw5.Calculator
open Hw5.MaybeBuilder

let (|CalculatorOperation|_|) arg =
    match arg with
    | Plus -> Some CalculatorOperation.Plus
    | Minus -> Some CalculatorOperation.Minus
    | Multiply -> Some CalculatorOperation.Multiply
    | Divide -> Some CalculatorOperation.Divide
    | _ -> None
    
let (|Double|_|) arg =
    match Double.TryParse(arg: string) with
    | true, double -> Some double
    | _ -> None

let isArgLengthSupported (args:string[]): Result<'a,'b> =
    match args.Length with
    | 3 -> Ok args
    | _ -> Error Message.WrongArgLength
    
[<System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage>]
let inline isOperationSupported (arg1, operation, arg2): Result<('a * CalculatorOperation * 'b), Message> =
    match operation with
    | CalculatorOperation operation -> Ok (arg1, operation, arg2)
    | _ -> Error Message.WrongArgFormatOperation

let parseArgs (args: string[]): Result<('a * string * 'b), Message> =
    match args[0] with
    | Double val1 ->
        match args[2] with
        | Double val2 -> Ok (val1, args[1], val2)
        | _ -> Error Message.WrongArgFormat
    | _ -> Error Message.WrongArgFormat

[<System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage>]
let inline isDividingByZero (arg1, operation, arg2): Result<('a * CalculatorOperation * 'b), Message> =
    if arg2 = 0.0 && operation = CalculatorOperation.Divide then Error Message.DivideByZero
    else Ok (arg1, operation, arg2)
    
let parseCalcArguments (args: string[]): Result<'a, 'b> =
    maybe {
        let! argLengthSupported = args |> isArgLengthSupported   
        let! argsParse = argLengthSupported |> parseArgs 
        let! operationParse = argsParse |> isOperationSupported 
        let! checkDividindByZero = operationParse |> isDividingByZero 
        return checkDividindByZero
    }